server/src/http/instance_data.rs

```diff
@@ -0,0 +1,122 @@
+use chrono::Duration;
+use reqwest::{StatusCode, Url};
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+use crate::cli::{CliArgs, EdgeMode};
+use crate::error::EdgeError;
+use crate::http::unleash_client::{new_reqwest_client, ClientMetaInformation, UnleashClient};
+use crate::metrics::edge_metrics::EdgeInstanceData;
+use prometheus::Registry;
+use tracing::{debug, trace, warn};
+
+#[derive(Debug, Clone)]
+pub struct InstanceDataSender {
+    pub unleash_client: Arc<UnleashClient>,
+    pub token: String,
+    pub base_path: String,
+}
+
+impl InstanceDataSender {
+    pub fn from_args(
+        args: CliArgs,
+        instance_data: Arc<EdgeInstanceData>,
+    ) -> Result<Option<Self>, EdgeError> {
+        match args.mode {
+            EdgeMode::Edge(edge_args) => {
+                let instance_id = instance_data.identifier.clone();
+                Ok(edge_args.tokens.first().map(|token| {
+                    let client_meta_information = ClientMetaInformation {
+                        app_name: args.app_name,
+                        instance_id,
+                    };
+                    let http_client = new_reqwest_client(
+                        edge_args.skip_ssl_verification,
+                        edge_args.client_identity.clone(),
+                        edge_args.upstream_certificate_file.clone(),
+                        Duration::seconds(edge_args.upstream_request_timeout),
+                        Duration::seconds(edge_args.upstream_socket_timeout),
+                        client_meta_information.clone(),
+                    )
+                    .expect("Could not construct reqwest client for posting observability data");
+                    let unleash_client = Url::parse(&edge_args.upstream_url.clone())
+                        .map(|url| {
+                            UnleashClient::from_url(
+                                url,
+                                args.token_header.token_header.clone(),
+                                http_client,
+                            )
+                        })
+                        .map(|c| {
+                            c.with_custom_client_headers(edge_args.custom_client_headers.clone())
+                        })
+                        .map(Arc::new)
+                        .map_err(|_| EdgeError::InvalidServerUrl(edge_args.upstream_url.clone()))
+                        .expect("Could not construct UnleashClient");
+                    Self {
+                        unleash_client,
+                        token: token.clone(),
+                        base_path: args.http.base_path.clone(),
+                    }
+                }))
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+pub async fn send_instance_data(
```

Comment:
Can we give this a different name? We now have two functions called the same thing doing different things

Reply:
yes